### PR TITLE
Added healthcheck

### PR DESCRIPTION
**Pull request by kngu9**

### CODE REVIEW

File: charm/kafka/reactive/kafka.py [hunk 1]
```diff
@@ -107,9 +109,11 @@ def waiting_for_certificates():
 )
 @when_not('kafka.started')
 def configure_kafka(zk):
+    if len(zk.zookeepers()) < 1:
+        hookenv.status_set('blocked', 'waiting for >1 zookeeper units')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: `>= 1`, or just `waiting for zookeeper`.

File: charm/kafka/reactive/kafka.py [hunk 2]
```diff
@@ -148,6 +149,10 @@ def configure_kafka_zookeepers(zk):
             data_changed('kafka.storage.log_dir', log_dir)))):
         return
 
+    if len(zks) < 1:
+        hookenv.status_set('blocked', 'waiting for >1 zookeeper units')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: `>= 1`, or just `waiting for zookeeper`.

File: charm/kafka/reactive/kafka.py [hunk 3]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
             remove_state('tls_client.ca_installed')
             set_state('kafka.ca.keystore.saved')
 
+@hook('upgrade-status')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: @hook('update-status'), see https://docs.jujucharms.com/2.5/en/reference-charm-hooks#update-status

File: charm/kafka/reactive/kafka.py [hunk 4]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
...
+        return
+
+    for i in range(3):
+        hookenv.status_set('maintenance', 'attempting to restart kafka, attempt: {}'.format(i))
+        kafka.restart()
+        if kafka.is_running():
+            return
+
+    hookenv.status_set('blocked', 'max times exceeded while trying to repair kafka')
+    # TODO: Figure out a repair mechanism for this or maybe just alert?
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: `update-status` will continue to fire every ~5 minutes. I think it's fine to just wait for the next go round, let this retry indefinitely.

File: charm/kafka/reactive/kafka.py [hunk 5]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
...
+    set_state('kafka.error')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: Don't need to set this state, if it's not used anywhere.

File: charm/kafka/reactive/kafka.py [hunk 6]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
...
+    if kafka.is_running():
+        return
+
+    for i in range(3):
+        hookenv.status_set('maintenance', 'attempting to restart kafka, attempt: {}'.format(i))
+        kafka.restart()
+        if kafka.is_running():
+            return
+
+    hookenv.status_set('blocked', 'max times exceeded while trying to repair kafka')
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: failed to start kafka; check syslog

File: charm/kafka/reactive/kafka.py [hunk 7]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
...
+@hook('upgrade-status')
+@when('kafka.started')
+def health_check():
+    # Health check will attempt to restart Kafka service, if the service is not running
+    kafka = Kafka()
+    if kafka.is_running():
+        return
+
+    for i in range(3):
+        hookenv.status_set('maintenance', 'attempting to restart kafka, attempt: {}'.format(i))
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py]: `i+1`, so it doesn't say `attempt: 0`?

File: charm/kafka/reactive/kafka.py [hunk 8]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
             remove_state('tls_client.ca_installed')
             set_state('kafka.ca.keystore.saved')
 
+@hook('upgrade-status')
+@when('kafka.started')
+def health_check():
+    # Health check will attempt to restart Kafka service, if the service is not running
+    kafka = Kafka()
+    if kafka.is_running():
+        return
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py:369]: Need to set status to active here, to clear prior statuses like mainteance, blocked.

File: charm/kafka/reactive/kafka.py [hunk 9]
```diff
@@ -355,6 +360,23 @@ def import_ca_crt_to_keystore():
...
+
+    for i in range(3):
+        hookenv.status_set('maintenance', 'attempting to restart kafka, attempt: {}'.format(i))
+        kafka.restart()
+        if kafka.is_running():
+            return
```
Comment (cmars) [inline charm/kafka/reactive/kafka.py:376]: Need to set status to active here, to clear prior statuses like mainteance, blocked.

**Review by cmars — CHANGES_REQUESTED** [review]
Hook name needs fixing. Probably needs some QA:

1. Deploy, manually systemctl stop the service, confirm it comes up
2. Deploy, manually systemctl stop the service, do something that prevents kafka from starting, confirm blocked, fix it, confirm it starts automatically.

File: charm/kafka/reactive/kafka.py (hunk 7 above)
Comment (kngu9) [reply charm/kafka/reactive/kafka.py]: yep!

**Review by cmars — APPROVED** [review]

File: charm/kafka/reactive/kafka.py (hunk 7 above)
Comment (cmars) [reply charm/kafka/reactive/kafka.py]: Still not changed...

**Pull request merged**